src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -706,6 +706,11 @@ private static void ReportBinaryOperatorError(ExpressionSyntax node, DiagnosticB
                     Error(diagnostics, ErrorCode.ERR_AmbigBinaryOpsOnDefault, node, operatorToken.Text);
                     return;
                 }
+                else if ((leftDefault && right.Type is TypeParameterSymbol) || (rightDefault && left.Type is TypeParameterSymbol))
```

Comment:
>else if ((leftDefault && right.Type is TypeParameterSymbol) || (rightDefault && left.Type is TypeParameterSymbol)) [](start = 16, length = 114)

This condition on itself doesn't check that we are dealing with an unconstrained type parameter. Is this checked elsewhere? Where?

Reply:
If the type parameter is constrained to a reference type, then we won't get here (copied spec below). In all other cases, we need to give this new diagnostic.

```
            // SPEC: (1) Both operands are a value of a type known to be a reference-type or the literal null.
            // SPEC:     Furthermore, an explicit reference conversion exists from the type of either
            // SPEC:     operand to the type of the other operand. Or:
            // SPEC: (2) One operand is a value of type T where T is a type-parameter and the other operand is 
            // SPEC:     the literal null. Furthermore T does not have the value type constraint.
            // SPEC: (3) One operand is the literal default and the other operand is a reference-type.
```

---
In reply to: [364496453](https://github.com/dotnet/roslyn/pull/40801#discussion_r364496453) [](ancestors = 364496453)

Reply:
From offline discussion, that check is done in `IsValidObjectEquality`